src/graph.rs

```diff
@@ -0,0 +1,74 @@
+use git2::{BranchType, Commit, Error, Oid, Repository};
+use std::collections::HashMap;
+
+#[allow(dead_code)]
+pub struct GitGraph {
+    pub repository: Repository,
+    pub commits: Vec<CommitInfo>,
+    pub indices: HashMap<Oid, usize>,
+}
+
+impl GitGraph {
+    pub fn new(path: &str) -> Result<Self, Error> {
+        let repository = Repository::open(path)?;
+        let mut walk = repository.revwalk()?;
+
+        walk.set_sorting(git2::Sort::TOPOLOGICAL)?;
+        walk.push_head()?;
+
+        let mut commits = Vec::new();
+        let mut indices = HashMap::new();
+        for (idx, oid) in walk.enumerate() {
+            let oid = oid?;
+            let commit = repository.find_commit(oid).unwrap();
+            commits.push(CommitInfo::new(&commit));
+            indices.insert(oid, idx);
+        }
+
+        let mut graph = GitGraph {
+            repository,
+            commits,
+            indices,
+        };
+        graph.assign_branches()?;
```

Comment:
It looks like `indices` is only used here? So maybe make it a temporary variable instead of putting it into `self`?

Reply:
It is currently used only in `assign_branches`, correct. Will probably need it when "iterating" along parents, but can introduce it again when needed.